enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,353 @@
+So only when the node plugin finishes the registration process with NRI and device plugin, the scheduler
+admits the application pod.
+
+Once a container under the pod requests and gets access to shared cpus, new environment variable named
+`OPENSHIFT_SHARED_CPUS=<CPU-IDs>` will be present under the container's environment variables.
+
+The mixed-cpu-node-plugin itself is essentially a pod that runs as a DaemonSet on the worker nodes.
+cluster-node-tuning-operator (NTO) will be modified to support the deployment of the DaemonSet and additional
+resources (ServiceAccount, RBAC resources, SecurityContextConstraint, etc.)
+NTO will also be responsible to watch, monitor and report the node-plugin's DaemonSet status.
+Additionally, since the plugin is part of OCP infrastructure pods, NTO annotates the pods with:
+```yaml
+target.workload.openshift.io/management: {
+  "effect":"PreferredDuringScheduling"
+}
+```
+See https://github.com/openshift/enhancements/blob/master/enhancements/workload-partitioning/management-workload-partitioning.md for details.
+
+Since both the components are related,
+mixed-cpu-node-plugin [source code](https://github.com/openshift-kni/mixed-cpu-node-plugin) would be vendor under NTO.
+By having different main.go + container entrypoint, NTO image can be also used as the
+mixed-cpu-node-plugin image. 
+This way saves the overhead of having a separate container image with all the maintenance cost that comes with it.
+
+In order to activate NRI in container-runtime, NTO creates MachineConfig to enable NRI in 
+the container-runtime configuration.
+
+The plugin is completely optional and is off by default.
+
+No changes are needed in the container-runtime, because NRI hooks already supported by the runtime
+and mixed-cpu-node-plugin implementing the NRI interface.
+This minimizes risks and helps avoid regression on
+a container-runtime component (CRI-O in OpenShift case).
+
+_"A picture is worth a thousand words"_
+![mixed-cpu-node-plugin](mixed-cpu-node-plugin.jpg)
+
+### API
+
+Extend PerformanceProfile with new `workloadHints` named `mixedCpus` in order to enable
+the feature.
+
+```yaml
+workloadHints:
+  mixedCpus: true
+```
+
+Specify the shared cpus via annotation:
+
+```yaml
+annotations:
+    mixed-cpu.openshift.io/shared-cpus: "2-3"
+```
+
+The CPU set value should follow the 
+[cpuset](https://github.com/openshift/cluster-node-tuning-operator/blob/master/docs/performanceprofile/performance_profile.md#cpuset) conventions.
+It defines a set of CPUs that can be allocated to guaranteed pods/containers that still require
+non-exclusive ones.
+
+The CPU set must not overlap with `spec.cpu.reserved` or `spec.cpu.isolated` and NTO
+should return an error in case it does.
+
+The reason for using an annotation and not new API field 
+is because there is a coming feature named shared-partition
+that uses similar convention i.e., shared cpus, and would be 
+enabled via NTO as well.
```

Comment:
I am confused about the relationship between the annotation here and the API field [here](https://github.com/openshift/enhancements/pull/1421). If they overlap, why not prioritize that so we only use one field. if they are different, why not use a different API field? I feel like mixing annotation and api fields here is confusing and is not very expressive for what is being developed.

Reply:
lol rtfm, you say it right below.

why not wait on this for shared-partitions, then?

Reply:
Hey @haircommander thank you for the review, appreciate it.
the shared-partition feature doesn't have a clear timeline yet, while we're striving to add this feature during 4.14/4.15 timeframe